features/192-asset-gear/index.md

```diff
@@ -0,0 +1,495 @@
+{
+  "type": "object",
+  "properties": {
+    "asset_id": { "type": "string" },
+    "url": { "type": "string" },
+    "method": { "type": "string" },
+    "headers": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "name": { "type": "string" },
+          "value": { "type": "string" }
+        }
+      }
+    }
+  },
+  "required": ["asset_id", "url", "method", "headers"]
+}
+```
+
+- `asset_id`: The asset ID. It becomes valid when the asset is uploaded.
+- `url`: The presigned URL to upload the asset.
+- `method`: The method to be used in the upload request.
+- `headers`: The headers to be included in the upload request.
+
+##### Response Example
+
+```json
+{
+  "asset_id": "myimage.png",
+  "url": "https://storage.skygearapps.com/bucket/myappid/myimage.png?...",
+  "method": "PUT",
+  "headers": [
+    { "name": "content-type", "value": "image/png" },
+    { "name": "content-md5", "value": "Zum77CZrrrGRDM18nlplig==" },
+    { "name": "content-length", "value": "26419" }
+  ]
+}
+```
+
+#### Specification
+
+1. Validate `content_type`.
+1. Let `name` be a randomly generated string.
+1. Let `ext` be the file extension derived from `content_type`.
+1. If `ext` is found, append it to `name`.
+1. Let `asset_id` be `/<app-id>/<name>`.
+1. Presign `asset_id` with `content_type`, `content_length`, `content_md5` as headers.
+1. Return the presigned request.
+
+### POST /_asset/sign
+
+#### Description
+
+This endpoint signs the given assets.
+
+This endpoint requires Master Key.
+
+#### Request JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "assets": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "asset_id": { "type": "string" }
+        },
+        "required": ["asset_id"]
+      }
+    }
+  },
+  "required": ["assets"]
+}
+```
+
+- `asset_id`: The asset ID.
+
+##### Request Example
+
+```json
+{
+  "assets": [
+    { "asset_id": "myimage.png" }
+  ]
+}
+```
+
+#### Response JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "assets": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "asset_id": { "type": "string" },
+          "url": { "type": "string" }
+        },
+        "required": ["asset_id", "url"]
+      }
+    }
+  },
+  "required": ["assets"]
+}
+```
+
+- `url`: The signed URL.
+
+##### Response Example
+
+```json
+{
+  "assets": [
+    {
+      "asset_id": "myimage.png",
+      "url": "https://myappname.skygearapps.com/_asset/get/myimage.png?..."
+    }
+  ]
+}
+```
+
+#### Specification
+
+1. Sign each `asset_id` in `assets`.
+1. Return the result in the original order.
+
+### GET /_asset/get/<asset_id>
+
+#### Request Query String
+
+The asset is eligible for image processing if all of the following hold
+
+- The asset is less than or equal to 20MiB.
+- The detected file type using libmagic is png, jpeg or webp.
```

Comment:
I think this would impose a hard dependency on libmagic on implementation. Could we just say 'the file is a valid png/jpeg/webp'?

Reply:
Let me remove `libmagic` from the spec. It is an implementation detail.